server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -204,7 +205,16 @@ private MappedFieldType getKeyedJsonField(String field) {
 
     @Override
     public Iterator<MappedFieldType> iterator() {
-        return fullNameToFieldType.values().iterator();
+        Iterator<MappedFieldType> concreteFieldTypes = fullNameToFieldType.values().iterator();
+
+        if (fullNameToJsonMapper.isEmpty()) {
+            return concreteFieldTypes;
+        } else {
+            Iterator<MappedFieldType> keyedJsonFieldTypes = fullNameToJsonMapper.values().stream()
```

Comment:
A note for future reference: the other option would have been to expose the field type by adding a 'keyed' field mapper in `JsonFieldMapper#iterator`. I didn't go down that path because this would make `my_json_field._keyed` a valid field to search on, which we want to avoid.